src/constraint/spec.ts

```diff
@@ -719,12 +719,27 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
         const xEncQ = specM.getEncodingQueryByChannel('x');
         const yEncQ = specM.getEncodingQueryByChannel('y');
 
-        // TODO(#186): take non-spatial channel into account
         if ((!isFieldQuery(xEncQ) || isDiscrete(xEncQ)) &&
-          (!isFieldQuery(yEncQ) || isDiscrete(yEncQ)) &&
-          !specM.isAggregate() // TODO: refactor based on statistics
-        ) {
-          return false;
+            (!isFieldQuery(yEncQ) || isDiscrete(yEncQ))) {
+          if (!specM.isAggregate()) {
+            return false;
+          } else {
+            return every(specM.getEncodings(), (encQ) => {
+              let channel = encQ.channel;
+
+              if (channel === Channel.COLOR || channel === Channel.DETAIL ||
+                  channel === Channel.SHAPE || channel === Channel.SIZE ||
+                  channel === Channel.OPACITY) {
+                if (encQ['aggregate']) {
```

Comment:
`['aggregate']` is a hack to get around type system.  Instead, please use the new `isFieldQuery` or `isAutoCountQuery` type guard functions.